torch/fx/passes/backends/cudagraphs.py

```diff
@@ -0,0 +1,53 @@
+import torch
+from torch.fx.passes.infra.partitioner import CapabilityBasedPartitioner
+from torch.fx.passes.operator_support import OperatorSupport
+from torch.fx.passes.tools_common import CALLABLE_NODE_OPS
+from torch.fx.passes.fake_tensor_prop import FakeTensorProp
+from torch.utils._pytree import tree_map
+
+import operator
+
+class CudaGraphsSupport(OperatorSupport):
+    # TODO: why is submodules passed here
+    def is_node_supported(self, submodules, node: torch.fx.Node) -> bool:
+        if node.op not in CALLABLE_NODE_OPS:
+            return False
+
+        if node.target in [torch.ops.aten.embedding_dense_backward.default]:
+            return False
+
+        if node.target in [operator.getitem]:
+            return True
+
+        found_not_cuda = False
+
+        def find_not_cuda(t):
+            nonlocal found_not_cuda
+            if isinstance(t, torch.Tensor) and t.device.type != 'cuda':
+                found_not_cuda = True
+
+        for n in node.all_input_nodes:
+            tree_map(find_not_cuda, n.meta['fake_result'])
+
+        tree_map(find_not_cuda, node.meta['fake_result'])
+
+        # NB: factory function is accounted for because the result would be
+        # cpu or cuda
+
+        return not found_not_cuda
+
+def partition_cudagraphs(gm, inputs):
+    """
+    Partition an FX graph into sub-GraphModules that can be validly run under
+    CUDA graphs.  For a subgraph to be runnable under CUDA, all of the operations
+    must involve CUDA tensors only/
+    """
+
+    FakeTensorProp(gm).propagate(*inputs)
+    supported_ops = CudaGraphsSupport()
+    # TODO: single node partition may be wrong due to the pessimization
```

Comment:
what would it mean if you disallowed single-node partitioning?  Is the 'good' case for single-node that there are 2 big cuda graphs surrounding a single-node cpu island?  but there is no prevention for having many small islands or even cuda-graph islands?

Reply:
A single node partition would be us trying to CUDA graph a single operator call on CUDA arguments, like at::add. This could still be profitable if the kernel has a lot of CPU overhead (which is the case with, e.g., TensorIterator), but this is counterbalanced by the need to call into and out of the CUDA graph.

The single node partition does NOT refer to CPU islands, those can be arbitrarily small; we're just talking about the size of the fusion groups